templates/nodeovc/nodeovc.py

```diff
@@ -0,0 +1,248 @@
+
+        return key         
+
+    @property
+    def ovc(self):
+        """
+        An ovc connection instance
+        """
+        if self._ovc is not None:
+            return self._ovc
+
+        self._ovc = j.clients.openvcloud.get(self.data['openvcloud'])
+        return self._ovc
+
+    @property
+    def space(self):
+        return self.ovc.space_get(accountName=None,
+                                 spaceName=self.data['openvcloud'])
+
+    @property
+    def machine(self):
+        if self._machine:
+            return self._machine
+        return self.space.machines.get(self.data['name'])
+
+    def install(self):
+        machine = self.machine
+        if not machine:
+            machine = self._machine_create()
+
+        self._configure_ports()
+
+        # Get data from the vm
+        ip_private, vm_info = machine.machineip_get()
+        self.data['sshLogin'] = vm_info['accounts'][0]['login']
+        self.data['sshPassword']= vm_info['accounts'][0]['password']
+        self.data['ipPrivate'] = ip_private
+        self.data['ipPublic'] = machine.space.model['publicipaddress']
+        self.data['machineId'] = machine.id
+
+        self.state.set('actions', 'install', 'ok')
+        self.save()
+
+    def uninstall(self):
+        # check if the machine is in the space
+        machine = self.machine
+        if machine:
+            self.machine.delete()
+        else:
+            self.err_log_machine_not_found()
+
+    def _machine_create(self):
+        """ Create a new machine """
+
+        self._machine =  self.space.machine_create(
+            name=self.data['name'],
+            sshkeyname= self.sshkey,
+            image=self.data['osImage'],
+            disksize=self.data['bootDiskSize'],
+            datadisks=self.data['disks'],
+            sizeId=self.data['sizeId'],
+            )
+        return self._machine
+
+    def _configure_ports(self):
+        """ Configure portforwards """
+
+        machine = self.machine
+        port_forwards = self.data['ports']
+        
+        # get list of existing ports at the vm
+        existent_ports = [port['publicPort'] for port in machine.portforwards]
+
+        # list of requested ports
+        requested_ports = [port['destination'] for port in port_forwards]
+
+        # check if port 22 is already created or requested
+        ssh_present = ('22' in existent_ports) or ('22' in requested_ports)
+
+        # if port 22 is not created, add to requested
+        if not ssh_present:
+            ssh_port = {'source':'22', 'destination': '22'}
+            port_forwards.append(ssh_port)
+
+        # delete ports that are not requested
+        for port in existent_ports:
+            if port != '22' or port not in requested_ports:
+                machine.portforward_delete(port)
+
+        # add requested ports
+        for port in port_forwards:
+            # check if ports do not exist yet
+            if port['destination'] not in existent_ports:
```

Comment:
shouldn't that be port['source'] ? cause `existent_ports` is built using the `port['publicPort']` and the publicPort is the source of a port forward, not the destination.
Or maybe the `existent_ports` should be built from the `port[privatePort]` ?

Reply:
I totally think that port forward management should be done on the VDC level. i.e a machine must call an action on the vdc to setup port forwarding, this way we have a central place to do port forwarding and also detect forwarding conflicts. and it will be easier to show all the forwarding rules this way.

Reply:
@muhamadazmy , `portforward_create` is a property of a `machine` object. Why and how to manage port forwarding on the `vdc` level, that does not do actions on VMs?

Reply:
Having the PF done on VDC level actually make sense cause that's where it belong. It's clear from the portal UI that the portforward are part of the vdc and not attached to a specific VM. 
That would greatly simplify the code both for the template and in the OVC client.

I know that ovc client doesn't present it like that now, but maybe we should change that